s3 upload fails on empty source

when trying to upload an empty file, we're getting this exception:
```
The future returned an exception of type: java.lang.RuntimeException, with message: No Responses.
ScalaTestFailureLocation: akka.stream.alpakka.s3.scaladsl.S3SinkSpec at (S3SinkSpec.scala:79)
org.scalatest.exceptions.TestFailedException: The future returned an exception of type: java.lang.RuntimeException, with message: No Responses.
	at org.scalatest.concurrent.Futures$FutureConcept.tryTryAgain$1(Futures.scala:531)
	at org.scalatest.concurrent.Futures$FutureConcept.futureValueImpl(Futures.scala:550)
	at org.scalatest.concurrent.Futures$FutureConcept.futureValueImpl$(Futures.scala:479)
	at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValueImpl(ScalaFutures.scala:275)
	at org.scalatest.concurrent.Futures$FutureConcept.futureValue(Futures.scala:476)
	at org.scalatest.concurrent.Futures$FutureConcept.futureValue$(Futures.scala:475)
	at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValue(ScalaFutures.scala:275)
	at akka.stream.alpakka.s3.scaladsl.S3SinkSpec.$anonfun$new$4(S3SinkSpec.scala:79)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at akka.stream.alpakka.s3.scaladsl.S3SinkSpec.withFixture(S3SinkSpec.scala:16)
	at org.scalatest.FlatSpecLike.invokeWithFixture$1(FlatSpecLike.scala:1680)
	at org.scalatest.FlatSpecLike.$anonfun$runTest$1(FlatSpecLike.scala:1692)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FlatSpecLike.runTest(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpecLike.runTest$(FlatSpecLike.scala:1674)
	at akka.stream.alpakka.s3.scaladsl.S3SinkSpec.runTest(S3SinkSpec.scala:16)
	at org.scalatest.FlatSpecLike.$anonfun$runTests$1(FlatSpecLike.scala:1750)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:396)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:373)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:410)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FlatSpecLike.runTests(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpecLike.runTests$(FlatSpecLike.scala:1749)
	at akka.stream.alpakka.s3.scaladsl.S3SinkSpec.runTests(S3SinkSpec.scala:16)
	at org.scalatest.Suite.run(Suite.scala:1147)
	at org.scalatest.Suite.run$(Suite.scala:1129)
	at akka.stream.alpakka.s3.scaladsl.S3SinkSpec.org$scalatest$FlatSpecLike$$super$run(S3SinkSpec.scala:16)
	at org.scalatest.FlatSpecLike.$anonfun$run$1(FlatSpecLike.scala:1795)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FlatSpecLike.run(FlatSpecLike.scala:1795)
	at org.scalatest.FlatSpecLike.run$(FlatSpecLike.scala:1793)
	at akka.stream.alpakka.s3.scaladsl.S3SinkSpec.org$scalatest$BeforeAndAfterAll$$super$run(S3SinkSpec.scala:16)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at akka.stream.alpakka.s3.scaladsl.S3SinkSpec.run(S3SinkSpec.scala:16)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1340)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1334)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1334)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:1031)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1500)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:131)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
Caused by: java.lang.RuntimeException: No Responses
	at akka.stream.alpakka.s3.impl.S3Stream.$anonfun$completionSink$2(S3Stream.scala:327)
	at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:304)
	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I was able to reproduce in a simple test case [here](https://github.com/eyalfa/alpakka/blob/empty-s3-upload-fails/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala#L67-L80),
as far as I understand s3 supports empty files/objects and this seems to be a library bug, as the library simply fails to produce any request for such cases.

I'm willing to take a stub at this and submit a pull request if the community agrees this is indeed a bug.

cc @jypma (as you seem to be the original author)

--- a/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
+++ b/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
@@ -16,11 +16,12 @@ import akka.http.scaladsl.model.StatusCodes.{NoContent, NotFound, OK}
 import akka.http.scaladsl.model._
 import akka.http.scaladsl.model.headers.ByteRange
 import akka.http.scaladsl.unmarshalling.{Unmarshal, Unmarshaller}
-import akka.stream.Materializer
+import akka.stream.{FlowShape, Materializer}
 import akka.stream.alpakka.s3.auth.{CredentialScope, Signer, SigningKey}
 import akka.stream.alpakka.s3.scaladsl.{ListBucketResultContents, ObjectMetadata}
 import akka.stream.alpakka.s3.{DiskBufferType, MemoryBufferType, S3Exception, S3Settings}
-import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
+import akka.stream.scaladsl.{Broadcast, Concat, Flow, GraphDSL, Keep, Sink, Source}
+import GraphDSL.Implicits._
 import akka.util.ByteString
 
 final case class S3Location(bucket: String, key: String)
@@ -267,7 +268,40 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
 
     val headers: S3Headers = S3Headers(sse.fold[Seq[HttpHeader]](Seq.empty) { _.headersFor(UploadPart) })
 
-    SplitAfterSize(chunkSize)(Flow.apply[ByteString])
+    import mat.executionContext
+
+    val extraSink = Sink
+      .fold(0){ (acc : Int, inp : ByteString) => acc + 1}
+      .mapMaterializedValue{ f =>
+        f.map{
+          case 0 => Some(ByteString.empty)
+          case _ => None
+        }
+      }
+
+    val conditionalAppendFl = Flow.fromGraph{
+      GraphDSL.create(extraSink){ implicit builder => extra =>
+        val extraSrc =
+          builder
+            .materializedValue
+            .mapAsync(1)(identity)
+            .collect{
+              case Some(b) => b
+            }
+        val bcast = builder.add(Broadcast[ByteString](2))
+        val concat = builder.add(Concat[ByteString]())
+
+        bcast.out(0) ~> extra
+
+        bcast.out(1) ~> concat.in(0)
+        extraSrc ~> concat.in(1)
+
+        FlowShape.of(bcast.in, concat.out)
+      }
+    }.mapMaterializedValue(_ => NotUsed)
+
+
+    SplitAfterSize(chunkSize)(/*Flow.apply[ByteString]*/conditionalAppendFl)
       .via(getChunkBuffer(chunkSize)) //creates the chunks
       .concatSubstreams
       .zipWith(requestInfo) {
@@ -301,16 +335,18 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
     val requestFlow = createRequests(s3Location, contentType, s3Headers, chunkSize, parallelism, sse)
 
     // The individual upload part requests are processed here
-    requestFlow.via(Http().superPool[(MultipartUpload, Int)]()).map {
-      case (Success(r), (upload, index)) =>
-        r.entity.dataBytes.runWith(Sink.ignore)
-        val etag = r.headers.find(_.lowercaseName() == "etag").map(_.value)
-        etag
-          .map((t) => SuccessfulUploadPart(upload, index, t))
-          .getOrElse(FailedUploadPart(upload, index, new RuntimeException("Cannot find etag")))
+    requestFlow
+      .via(Http().superPool[(MultipartUpload, Int)]())
+      .map {
+        case (Success(r), (upload, index)) =>
+          r.entity.dataBytes.runWith(Sink.ignore)
+          val etag = r.headers.find(_.lowercaseName() == "etag").map(_.value)
+          etag
+            .map((t) => SuccessfulUploadPart(upload, index, t))
+            .getOrElse(FailedUploadPart(upload, index, new RuntimeException(s"Cannot find etag in ${r}")))
 
-      case (Failure(e), (upload, index)) => FailedUploadPart(upload, index, e)
-    }
+        case (Failure(e), (upload, index)) => FailedUploadPart(upload, index, e)
+      }
   }
 
   private def completionSink(

--- a/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
+++ b/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
@@ -240,6 +240,8 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
       .zip(Source.fromIterator(() => Iterator.from(1)))
 
   //todo: this seems like a general purpose 'gadget', is there a better place for this kind of beasts?
+  val conditionalAppendFl = Flow[ByteString].orElse(Source.single(ByteString.empty))/*
+
   private def conditionalAppendFl: Flow[ByteString, ByteString, NotUsed.type] =
     Flow
       .fromGraph {
@@ -272,7 +274,7 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
           FlowShape.of(bcast.in, concat.out)
         }
       }
-      .mapMaterializedValue(_ => NotUsed)
+      .mapMaterializedValue(_ => NotUsed)*/
 
   private def createRequests(
       s3Location: S3Location,

--- a/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
+++ b/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
@@ -16,12 +16,11 @@ import akka.http.scaladsl.model.StatusCodes.{NoContent, NotFound, OK}
 import akka.http.scaladsl.model._
 import akka.http.scaladsl.model.headers.ByteRange
 import akka.http.scaladsl.unmarshalling.{Unmarshal, Unmarshaller}
-import akka.stream.{FlowShape, Materializer}
+import akka.stream.Materializer
 import akka.stream.alpakka.s3.auth.{CredentialScope, Signer, SigningKey}
 import akka.stream.alpakka.s3.scaladsl.{ListBucketResultContents, ObjectMetadata}
 import akka.stream.alpakka.s3.{DiskBufferType, MemoryBufferType, S3Exception, S3Settings}
-import akka.stream.scaladsl.{Broadcast, Concat, Flow, GraphDSL, Keep, Sink, Source}
-import GraphDSL.Implicits._
+import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
 import akka.util.ByteString
 
 final case class S3Location(bucket: String, key: String)
@@ -239,7 +238,7 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
       .mapConcat(r => Stream.continually(r))
       .zip(Source.fromIterator(() => Iterator.from(1)))
 
-  val conditionalAppendFl = Flow[ByteString].orElse(Source.single(ByteString.empty))
+  val atLeastOneByteString = Flow[ByteString].orElse(Source.single(ByteString.empty))
 
   private def createRequests(
       s3Location: S3Location,
@@ -270,7 +269,7 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
 
     val headers: S3Headers = S3Headers(sse.fold[Seq[HttpHeader]](Seq.empty) { _.headersFor(UploadPart) })
 
-    SplitAfterSize(chunkSize)(conditionalAppendFl)
+    SplitAfterSize(chunkSize)(atLeastOneByteString)
       .via(getChunkBuffer(chunkSize)) //creates the chunks
       .concatSubstreams
       .zipWith(requestInfo) {

--- a/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
+++ b/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
@@ -240,35 +240,39 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
       .zip(Source.fromIterator(() => Iterator.from(1)))
 
   //todo: this seems like a general purpose 'gadget', is there a better place for this kind of beasts?
-  private def conditionalAppendFl: Flow[ByteString, ByteString, NotUsed.type] = Flow.fromGraph{
-    import mat.executionContext
-    val extraSink = Sink
-      .fold(0){ (acc : Int, inp : ByteString) => acc + 1}
-      .mapMaterializedValue{ f =>
-        f.map{
-          case 0 => Some(ByteString.empty)
-          case _ => None
+  private def conditionalAppendFl: Flow[ByteString, ByteString, NotUsed.type] =
+    Flow
+      .fromGraph {
+        import mat.executionContext
+        val extraSink = Sink
+          .fold(0) { (acc: Int, inp: ByteString) =>
+            acc + 1
+          }
+          .mapMaterializedValue { f =>
+            f.map {
+              case 0 => Some(ByteString.empty)
+              case _ => None
+            }
+          }
+        GraphDSL.create(extraSink) { implicit builder => extra =>
+          val extraSrc =
+            builder.materializedValue
+              .mapAsync(1)(identity)
+              .collect {
+                case Some(b) => b
+              }
+          val bcast = builder.add(Broadcast[ByteString](2))
+          val concat = builder.add(Concat[ByteString]())
+
+          bcast.out(0) ~> extra
+
+          bcast.out(1) ~> concat.in(0)
+          extraSrc ~> concat.in(1)
+
+          FlowShape.of(bcast.in, concat.out)
         }
       }
-    GraphDSL.create(extraSink){ implicit builder => extra =>
-      val extraSrc =
-        builder
-          .materializedValue
-          .mapAsync(1)(identity)
-          .collect{
-            case Some(b) => b
-          }
-      val bcast = builder.add(Broadcast[ByteString](2))
-      val concat = builder.add(Concat[ByteString]())
-
-      bcast.out(0) ~> extra
-
-      bcast.out(1) ~> concat.in(0)
-      extraSrc ~> concat.in(1)
-
-      FlowShape.of(bcast.in, concat.out)
-    }
-  }.mapMaterializedValue(_ => NotUsed)
+      .mapMaterializedValue(_ => NotUsed)
 
   private def createRequests(
       s3Location: S3Location,

--- a/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
+++ b/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
@@ -239,42 +239,7 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
       .mapConcat(r => Stream.continually(r))
       .zip(Source.fromIterator(() => Iterator.from(1)))
 
-  //todo: this seems like a general purpose 'gadget', is there a better place for this kind of beasts?
-  val conditionalAppendFl = Flow[ByteString].orElse(Source.single(ByteString.empty)) /*
-
-  private def conditionalAppendFl: Flow[ByteString, ByteString, NotUsed.type] =
-    Flow
-      .fromGraph {
-        import mat.executionContext
-        val extraSink = Sink
-          .fold(0) { (acc: Int, inp: ByteString) =>
-            acc + 1
-          }
-          .mapMaterializedValue { f =>
-            f.map {
-              case 0 => Some(ByteString.empty)
-              case _ => None
-            }
-          }
-        GraphDSL.create(extraSink) { implicit builder => extra =>
-          val extraSrc =
-            builder.materializedValue
-              .mapAsync(1)(identity)
-              .collect {
-                case Some(b) => b
-              }
-          val bcast = builder.add(Broadcast[ByteString](2))
-          val concat = builder.add(Concat[ByteString]())
-
-          bcast.out(0) ~> extra
-
-          bcast.out(1) ~> concat.in(0)
-          extraSrc ~> concat.in(1)
-
-          FlowShape.of(bcast.in, concat.out)
-        }
-      }
-      .mapMaterializedValue(_ => NotUsed)*/
+  val conditionalAppendFl = Flow[ByteString].orElse(Source.single(ByteString.empty))
 
   private def createRequests(
       s3Location: S3Location,

--- a/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
+++ b/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
@@ -240,7 +240,7 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
       .zip(Source.fromIterator(() => Iterator.from(1)))
 
   //todo: this seems like a general purpose 'gadget', is there a better place for this kind of beasts?
-  val conditionalAppendFl = Flow[ByteString].orElse(Source.single(ByteString.empty))/*
+  val conditionalAppendFl = Flow[ByteString].orElse(Source.single(ByteString.empty)) /*
 
   private def conditionalAppendFl: Flow[ByteString, ByteString, NotUsed.type] =
     Flow

--- a/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
+++ b/s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala
@@ -239,6 +239,37 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
       .mapConcat(r => Stream.continually(r))
       .zip(Source.fromIterator(() => Iterator.from(1)))
 
+  //todo: this seems like a general purpose 'gadget', is there a better place for this kind of beasts?
+  private def conditionalAppendFl: Flow[ByteString, ByteString, NotUsed.type] = Flow.fromGraph{
+    import mat.executionContext
+    val extraSink = Sink
+      .fold(0){ (acc : Int, inp : ByteString) => acc + 1}
+      .mapMaterializedValue{ f =>
+        f.map{
+          case 0 => Some(ByteString.empty)
+          case _ => None
+        }
+      }
+    GraphDSL.create(extraSink){ implicit builder => extra =>
+      val extraSrc =
+        builder
+          .materializedValue
+          .mapAsync(1)(identity)
+          .collect{
+            case Some(b) => b
+          }
+      val bcast = builder.add(Broadcast[ByteString](2))
+      val concat = builder.add(Concat[ByteString]())
+
+      bcast.out(0) ~> extra
+
+      bcast.out(1) ~> concat.in(0)
+      extraSrc ~> concat.in(1)
+
+      FlowShape.of(bcast.in, concat.out)
+    }
+  }.mapMaterializedValue(_ => NotUsed)
+
   private def createRequests(
       s3Location: S3Location,
       contentType: ContentType,
@@ -268,40 +299,7 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
 
     val headers: S3Headers = S3Headers(sse.fold[Seq[HttpHeader]](Seq.empty) { _.headersFor(UploadPart) })
 
-    import mat.executionContext
-
-    val extraSink = Sink
-      .fold(0){ (acc : Int, inp : ByteString) => acc + 1}
-      .mapMaterializedValue{ f =>
-        f.map{
-          case 0 => Some(ByteString.empty)
-          case _ => None
-        }
-      }
-
-    val conditionalAppendFl = Flow.fromGraph{
-      GraphDSL.create(extraSink){ implicit builder => extra =>
-        val extraSrc =
-          builder
-            .materializedValue
-            .mapAsync(1)(identity)
-            .collect{
-              case Some(b) => b
-            }
-        val bcast = builder.add(Broadcast[ByteString](2))
-        val concat = builder.add(Concat[ByteString]())
-
-        bcast.out(0) ~> extra
-
-        bcast.out(1) ~> concat.in(0)
-        extraSrc ~> concat.in(1)
-
-        FlowShape.of(bcast.in, concat.out)
-      }
-    }.mapMaterializedValue(_ => NotUsed)
-
-
-    SplitAfterSize(chunkSize)(/*Flow.apply[ByteString]*/conditionalAppendFl)
+    SplitAfterSize(chunkSize)(conditionalAppendFl)
       .via(getChunkBuffer(chunkSize)) //creates the chunks
       .concatSubstreams
       .zipWith(requestInfo) {

--- a/s3/src/test/scala/akka/stream/alpakka/s3/impl/S3StreamSpec.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/impl/S3StreamSpec.scala
@@ -4,21 +4,27 @@
 
 package akka.stream.alpakka.s3.impl
 
+import akka.NotUsed
 import akka.actor.ActorSystem
 import akka.http.scaladsl.model.HttpRequest
 import akka.http.scaladsl.model.headers.ByteRange
 import akka.stream.alpakka.s3.{MemoryBufferType, S3Settings}
+import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
 import akka.stream.{ActorMaterializer, ActorMaterializerSettings}
 import akka.testkit.TestKit
+import akka.util.ByteString
 import com.amazonaws.auth.{AWSStaticCredentialsProvider, BasicAWSCredentials}
 import com.amazonaws.regions.AwsRegionProvider
+import org.scalatest.concurrent.{IntegrationPatience, ScalaFutures}
 import org.scalatest.{FlatSpecLike, Matchers, PrivateMethodTester}
 
 class S3StreamSpec(_system: ActorSystem)
     extends TestKit(_system)
     with FlatSpecLike
     with Matchers
-    with PrivateMethodTester {
+    with PrivateMethodTester
+    with ScalaFutures
+    with IntegrationPatience {
 
   import HttpRequests._
 
@@ -73,4 +79,45 @@ class S3StreamSpec(_system: ActorSystem)
     result.headers.seq.exists(_.lowercaseName() == "range")
 
   }
+
+  it should "properly handle input streams, even when empty" in  {
+    val credentialsProvider =
+      new AWSStaticCredentialsProvider(
+        new BasicAWSCredentials(
+          "test-Id",
+          "test-key"
+        )
+      )
+    val regionProvider =
+      new AwsRegionProvider {
+        def getRegion: String = "us-east-1"
+      }
+    implicit val settings = new S3Settings(MemoryBufferType, None, credentialsProvider, regionProvider, false, None)
+    val s3stream = new S3Stream(settings)
+
+    def nonEmptySrc = Source.repeat(ByteString("hello world"))
+    val conditionallyAppended = s3stream invokePrivate PrivateMethod[Flow[ByteString, ByteString, NotUsed]]('conditionalAppendFl)()
+
+    nonEmptySrc
+      .take(1)
+      .via(conditionallyAppended)
+      .toMat(Sink.seq[ByteString])(Keep.right)
+      .run()
+      .futureValue should equal (Seq(ByteString("hello world")))
+
+    nonEmptySrc
+      .take(10)
+      .via(conditionallyAppended)
+      .toMat(Sink.seq[ByteString])(Keep.right)
+      .run()
+      .futureValue should equal (Seq.fill(10)(ByteString("hello world")))
+
+    nonEmptySrc
+      .take(0)
+      .via(conditionallyAppended)
+      .toMat(Sink.seq[ByteString])(Keep.right)
+      .run()
+      .futureValue should equal (Seq(ByteString.empty))
+  }
+
 }

--- a/s3/src/test/scala/akka/stream/alpakka/s3/impl/S3StreamSpec.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/impl/S3StreamSpec.scala
@@ -96,27 +96,24 @@ class S3StreamSpec(_system: ActorSystem)
     val s3stream = new S3Stream(settings)
 
     def nonEmptySrc = Source.repeat(ByteString("hello world"))
-    val conditionallyAppended = s3stream invokePrivate PrivateMethod[Flow[ByteString, ByteString, NotUsed]](
-      'conditionalAppendFl
-    )()
 
     nonEmptySrc
       .take(1)
-      .via(conditionallyAppended)
+      .via(s3stream.atLeastOneByteString)
       .toMat(Sink.seq[ByteString])(Keep.right)
       .run()
       .futureValue should equal(Seq(ByteString("hello world")))
 
     nonEmptySrc
       .take(10)
-      .via(conditionallyAppended)
+      .via(s3stream.atLeastOneByteString)
       .toMat(Sink.seq[ByteString])(Keep.right)
       .run()
       .futureValue should equal(Seq.fill(10)(ByteString("hello world")))
 
     nonEmptySrc
       .take(0)
-      .via(conditionallyAppended)
+      .via(s3stream.atLeastOneByteString)
       .toMat(Sink.seq[ByteString])(Keep.right)
       .run()
       .futureValue should equal(Seq(ByteString.empty))

--- a/s3/src/test/scala/akka/stream/alpakka/s3/impl/S3StreamSpec.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/impl/S3StreamSpec.scala
@@ -80,7 +80,7 @@ class S3StreamSpec(_system: ActorSystem)
 
   }
 
-  it should "properly handle input streams, even when empty" in  {
+  it should "properly handle input streams, even when empty" in {
     val credentialsProvider =
       new AWSStaticCredentialsProvider(
         new BasicAWSCredentials(
@@ -96,28 +96,30 @@ class S3StreamSpec(_system: ActorSystem)
     val s3stream = new S3Stream(settings)
 
     def nonEmptySrc = Source.repeat(ByteString("hello world"))
-    val conditionallyAppended = s3stream invokePrivate PrivateMethod[Flow[ByteString, ByteString, NotUsed]]('conditionalAppendFl)()
+    val conditionallyAppended = s3stream invokePrivate PrivateMethod[Flow[ByteString, ByteString, NotUsed]](
+      'conditionalAppendFl
+    )()
 
     nonEmptySrc
       .take(1)
       .via(conditionallyAppended)
       .toMat(Sink.seq[ByteString])(Keep.right)
       .run()
-      .futureValue should equal (Seq(ByteString("hello world")))
+      .futureValue should equal(Seq(ByteString("hello world")))
 
     nonEmptySrc
       .take(10)
       .via(conditionallyAppended)
       .toMat(Sink.seq[ByteString])(Keep.right)
       .run()
-      .futureValue should equal (Seq.fill(10)(ByteString("hello world")))
+      .futureValue should equal(Seq.fill(10)(ByteString("hello world")))
 
     nonEmptySrc
       .take(0)
       .via(conditionallyAppended)
       .toMat(Sink.seq[ByteString])(Keep.right)
       .run()
-      .futureValue should equal (Seq(ByteString.empty))
+      .futureValue should equal(Seq(ByteString.empty))
   }
 
 }

--- a/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
@@ -53,8 +53,6 @@ class S3SinkSpec extends S3WireMockBase with S3ClientIntegrationSpec {
     result.futureValue shouldBe MultipartUploadResult(url, bucket, bucketKey, etag)
   }
 
-
-
   it should "upload a stream of bytes to S3 with custom headers" in {
 
     //mockUpload()
@@ -79,7 +77,6 @@ class S3SinkSpec extends S3WireMockBase with S3ClientIntegrationSpec {
 
     result.failed.futureValue.getMessage shouldBe "No key found"
   }
-
 
   override protected def afterAll(): Unit = {
     super.afterAll()

--- a/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
@@ -26,6 +26,20 @@ class S3SinkSpec extends S3WireMockBase with S3ClientIntegrationSpec {
   val settings = new S3Settings(MemoryBufferType, proxy, awsCredentialsProvider, regionProvider, false, None)
   val s3Client = new S3Client(settings)(system, materializer)
 
+  it should "succeed uploading an empty file" in {
+    mockUpload(expectedBody = "")
+
+    //#upload
+    val s3Sink: Sink[ByteString, Future[MultipartUploadResult]] = s3Client.multipartUpload(bucket, bucketKey)
+    //#upload
+
+    val src = Source.empty[ByteString]
+
+    val result: Future[MultipartUploadResult] = src.runWith(s3Sink)
+
+    result.futureValue shouldBe MultipartUploadResult(url, bucket, bucketKey, etag)
+  }
+
   "S3Sink" should "upload a stream of bytes to S3" in {
 
     mockUpload()
@@ -39,9 +53,11 @@ class S3SinkSpec extends S3WireMockBase with S3ClientIntegrationSpec {
     result.futureValue shouldBe MultipartUploadResult(url, bucket, bucketKey, etag)
   }
 
+
+
   it should "upload a stream of bytes to S3 with custom headers" in {
 
-    mockUpload()
+    //mockUpload()
 
     //#upload
     val s3Sink: Sink[ByteString, Future[MultipartUploadResult]] =
@@ -64,20 +80,6 @@ class S3SinkSpec extends S3WireMockBase with S3ClientIntegrationSpec {
     result.failed.futureValue.getMessage shouldBe "No key found"
   }
 
-  it should "succeed uploading an empty file" in {
-    mockUpload()
-
-    //#upload
-    val s3Sink: Sink[ByteString, Future[MultipartUploadResult]] =
-      s3Client.multipartUploadWithHeaders(bucket, bucketKey, s3Headers = Some(S3Headers(ServerSideEncryption.AES256)))
-    //#upload
-
-    val src = Source.empty[ByteString]
-
-    val result: Future[MultipartUploadResult] = src.runWith(s3Sink)
-
-    result.futureValue shouldBe MultipartUploadResult(url, bucket, bucketKey, etag)
-  }
 
   override protected def afterAll(): Unit = {
     super.afterAll()

--- a/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
@@ -55,7 +55,7 @@ class S3SinkSpec extends S3WireMockBase with S3ClientIntegrationSpec {
 
   it should "upload a stream of bytes to S3 with custom headers" in {
 
-    //mockUpload()
+    mockUpload()
 
     //#upload
     val s3Sink: Sink[ByteString, Future[MultipartUploadResult]] =

--- a/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3SinkSpec.scala
@@ -64,6 +64,21 @@ class S3SinkSpec extends S3WireMockBase with S3ClientIntegrationSpec {
     result.failed.futureValue.getMessage shouldBe "No key found"
   }
 
+  it should "succeed uploading an empty file" in {
+    mockUpload()
+
+    //#upload
+    val s3Sink: Sink[ByteString, Future[MultipartUploadResult]] =
+      s3Client.multipartUploadWithHeaders(bucket, bucketKey, s3Headers = Some(S3Headers(ServerSideEncryption.AES256)))
+    //#upload
+
+    val src = Source.empty[ByteString]
+
+    val result: Future[MultipartUploadResult] = src.runWith(s3Sink)
+
+    result.futureValue shouldBe MultipartUploadResult(url, bucket, bucketKey, etag)
+  }
+
   override protected def afterAll(): Unit = {
     super.afterAll()
     stopWireMockServer()

--- a/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3WireMockBase.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3WireMockBase.scala
@@ -165,7 +165,8 @@ abstract class S3WireMockBase(_system: ActorSystem, _wireMockServer: WireMockSer
         )
       )
 
-  def mockUpload(): Unit = {
+  def mockUpload() : Unit = mockUpload(body)
+  def mockUpload(expectedBody : String): Unit = {
     mock
       .register(
         post(urlEqualTo(s"/$bucketKey?uploads")).willReturn(
@@ -184,7 +185,7 @@ abstract class S3WireMockBase(_system: ActorSystem, _wireMockServer: WireMockSer
 
     mock.register(
       put(urlEqualTo(s"/$bucketKey?partNumber=1&uploadId=$uploadId"))
-        .withRequestBody(matching(body))
+        .withRequestBody(matching(expectedBody))
         .willReturn(
           aResponse()
             .withStatus(200)

--- a/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3WireMockBase.scala
+++ b/s3/src/test/scala/akka/stream/alpakka/s3/scaladsl/S3WireMockBase.scala
@@ -165,8 +165,8 @@ abstract class S3WireMockBase(_system: ActorSystem, _wireMockServer: WireMockSer
         )
       )
 
-  def mockUpload() : Unit = mockUpload(body)
-  def mockUpload(expectedBody : String): Unit = {
+  def mockUpload(): Unit = mockUpload(body)
+  def mockUpload(expectedBody: String): Unit = {
     mock
       .register(
         post(urlEqualTo(s"/$bucketKey?uploads")).willReturn(

